File: 200906_I_Claudius/convert.rb
#!/usr/bin/ruby -w

NUMERALS = {1000 => 'M',
             900 => 'CM', 500 => 'D', 400 => 'CD', 100 => 'C',
              90 => 'XC',  50 => 'L',  40 => 'XL',  10 => 'X',
               9 => 'IX',   5 => 'V',   4 => 'IV',   1 => 'I'}

number = ARGV[0] =~ /^\d+$/ && (1..3999).member?(ARGV[0].to_i) ? ARGV[0].to_i : nil

if nil == number
  puts "The first command line argument (#{number}) must be a number between 1 and 3999 inclusive"
  exit 1
end

result = NUMERALS.keys.sort {|a, b| b <=> a}.inject ""  do |s, i|
   while number >= i do
     number -= i
     s += NUMERALS[i]
   end
   s
end

puts result

File: 2009/06_I_Claudius/convert.clj
(def numerals 
    {1000 'M 900 'CM 500 'D 400 'CD 100 'C 90 'XC 
     50 'L 40 'XL 10 'X 9 'IX 5 'V 4 'IV 1 'I }
) 

(defn error [message]
    (println message)
    (System/exit 1)
)

(defn get-command-line-arg [] 
    (try (let [i (Integer/parseInt (first *command-line-args*))] 
          (if (< i 1) (error "Number cannot be smaller than one"))
          (if (> i 3999) (error "Number must be smaller than 4000"))
          i)
    (catch NumberFormatException e (error "Not a number")))
)

(defn get-highest-symbol [number]
    (first (sort (fn [x y] (- y x)) (filter #(>= number %) (keys numerals))))
)

(defn convert [number]
    (if (> number 0)
        (str (get numerals (get-highest-symbol number))
             (convert (- number (get-highest-symbol number))))
    )
)

(println (convert (get-command-line-arg)))
